src/FileCacheManager.php

```diff
@@ -32,144 +35,76 @@
+        $this->readCache();
     }
 
-    private function calcHash($content)
+    public function __destruct()
     {
-        return crc32($content);
+        $this->writeCache();
     }
 
-    private function isCacheAvailable()
+    private function readCache()
     {
-        static $result;
+        $cache = $this->handler->read();
 
-        if (null === $result) {
-            $result = $this->isEnabled && (ToolInfo::isInstalledAsPhar() || ToolInfo::isInstalledByComposer());
+        if (!$cache || !$this->signature->equals($cache->signature())) {
+            $cache = new Cache($this->signature);
         }
 
-        return $result;
+        $this->cache = $cache;
     }
 
-    private function isCacheStale($php, $version, $linting, $rules)
+    private function writeCache()
     {
-        if (!$this->isCacheAvailable()) {
-            return true;
-        }
-
-        return PHP_VERSION !== $php || ToolInfo::getVersion() !== $version || $this->linting !== $linting || $this->rules !== $rules;
+        $this->handler->write($this->cache);
     }
 
-    private function readFromFile()
+    public function needFixing($file, $fileContent)
     {
-        if (!$this->isCacheAvailable()) {
-            return;
-        }
-
-        if (!file_exists($this->cacheFile)) {
-            return;
-        }
-
-        $content = file_get_contents($this->cacheFile);
-        $data = @unserialize($content);
-
-        // ignore corrupted serialized data
-        if (null === $data) {
-            return;
-        }
+        $file = $this->getRelativePathname($file);
 
-        if (!isset($data['php'], $data['version'], $data['linting'], $data['rules'])) {
-            return;
+        if ($this->cache->get($file) !== $this->calcHash($fileContent)) {
+            return true;
         }
 
-        // Set hashes only if the cache is fresh, otherwise we need to parse all files
-        if (!$this->isCacheStale($data['php'], $data['version'], $data['linting'], $data['rules'])) {
-            $this->oldHashes = $data['hashes'];
-            $this->newHashes = $this->oldHashes;
-        }
+        return false;
     }
 
-    private function saveToFile()
+    public function setFile($file, $fileContent)
     {
-        if (!$this->isCacheAvailable()) {
-            return;
-        }
+        $file = $this->getRelativePathname($file);
 
-        $data = serialize(
-            array(
-                'php' => PHP_VERSION,
-                'version' => ToolInfo::getVersion(),
-                'linting' => $this->linting,
-                'rules' => $this->rules,
-                'hashes' => $this->newHashes,
-            )
-        );
-
-        if (false === @file_put_contents($this->cacheFile, $data, LOCK_EX)) {
-            throw new IOException(sprintf('Failed to write file "%s".', $this->cacheFile), 0, null, $this->cacheFile);
-        }
+        $this->cache->set($file, $this->calcHash($fileContent));
```

Comment:
@keradus @stof

The fix for the original issue would then probably be to

* inject `$isDryRun` into `FileCacheManager`
* adjust `setFile()`:

    ```php
    public function setFile($file, $fileContent)
    {
        $file = $this->getRelativePathname($file);

        if ($this->isDryRun && $this->cache->get($file) !== $this->calcHash($fileContent)) {
            $this->cache->clear($file);
            
            return;
        }

        $this->cache->set($file, $this->calcHash($fileContent));
    }
    ```